feat!: fiat-shamir AVM vk

lucasxia01 opened the pull request:
Passes the vk to the prover and hardcodes the vk and vk hash in the recursive verifier. Adds logic to fiat-shamir VK hash in AVM, but stops short of doing it because padding makes things inconsistent in the recursive verifier. Also fixes NegativeBadPublicInputs test.

jeanmon commented on barretenberg/cpp/src/barretenberg/vm2/constraining/flavor.hpp:
```diff
@@ -268,28 +268,9 @@ class AvmFlavor {
 
-        /**
-         * @brief Serialize verification key to field elements
-         *
```
Was this removed on purpose?

jeanmon commented on barretenberg/cpp/src/barretenberg/vm2/constraining/verifier.cpp:
```diff
@@ -55,13 +55,11 @@ bool AvmVerifier::verify_proof(const HonkProof& proof, const std::vector<std::ve
+    vinfo("AVM vk hash in verifier: ", vkey_hash);
 
-    const auto circuit_size = transcript->template receive_from_prover<uint32_t>("circuit_size");
-    if (circuit_size != key->circuit_size) {
```
If we remove this sanity check, maybe the circuit_size should be removed from transcript?
@lucasxia01 What is the current approach of crypto team for the other proving system? Are you removing the circuit size from transcript?

lucasxia01 commented on barretenberg/cpp/src/barretenberg/vm2/constraining/prover.cpp:
```diff
@@ -36,9 +39,10 @@ AvmProver::AvmProver(std::shared_ptr<Flavor::ProvingKey> input_key, const PCSCom
-    const auto circuit_size = static_cast<uint32_t>(key->circuit_size);
-
-    transcript->send_to_verifier("circuit_size", circuit_size);
+    // TODO(#15892): Fiat-shamir the vk hash by uncommenting the line below.
```
could probably create a separate issue for FS as this issue is for fixing the VK in-circuit

ledwards2225 replied:
Seems like they would reasonably handled simultaneously no?

lucasxia01 replied:
yeah

lucasxia01 commented on barretenberg/cpp/src/barretenberg/vm2/constraining/recursion/goblin_avm_recursive_verifier.hpp:
```diff
@@ -135,6 +141,7 @@ class AvmGoblinRecursiveVerifier {
         // Fix the inner mega vk and vk hash to be constants in the outer circuit.
         mega_vk_and_hash->vk->fix_witness();
         mega_vk_and_hash->hash.fix_witness();
+        // TODO(#16352): Figure out how not to add hashing gates to mega recursive verifier.
```
outlined in the issue itself but we do extra hashing to check vk and vk hash consistency in the mega recursive verifier and its not needed in this context because we are hardcoded them into the circuit

lucasxia01 commented on barretenberg/cpp/src/barretenberg/vm2/constraining/flavor.hpp:
```diff
@@ -268,28 +268,9 @@ class AvmFlavor {
-
-        /**
-         * @brief Adds the verification key hash to the transcript and returns the hash.
-         * @details Needed to make sure the Origin Tag system works. See the base class function for
```
its not implemented so the details don't matter

ledwards2225 commented on barretenberg/cpp/src/barretenberg/vm2/constraining/recursion/goblin_avm_recursive_verifier.hpp:
```diff
@@ -204,7 +211,13 @@ class AvmGoblinRecursiveVerifier {
+        // buffer
         std::vector<FF> key_fields = convert_stdlib_ultra_to_stdlib_mega(outer_key_fields);
+        FF mega_vk_hash = convert_stdlib_ultra_to_stdlib_mega({ vk_hash })[0];
+        // TODO(#15892): We call fix_witness on the vk hash twice here because we hash it here and also for
```
Not sure I understand what is meant by this comment

lucasxia01 replied:
yea, this is a little ugly. Here, we are putting the AVM vk hash in a hash buffer to check that both the inner and outer AVM recursive verifier circuits are actually verifying the same AVM proof/vk. In the actual verify_proof call a few lines below, we are also fiat-shamiring the AVM vk hash. We need to connect these two, and one way is to just fix_witness on both so the circuit shows that they're the same value.

ledwards2225 commented on barretenberg/cpp/src/barretenberg/vm2/constraining/recursion/recursive_verifier.hpp:
```diff
@@ -35,8 +33,9 @@ class AvmRecursiveVerifier {
-    std::shared_ptr<VerificationKey> key;
     Builder& builder;
+    std::shared_ptr<VerificationKey> key;
+    FF vk_hash;
```
does this need to be a member var?

lucasxia01 replied:
probably not. I think it can be moved to the verify_proof call

ledwards2225 reviewed (approved):
LGTM, appreciate you setting this up

jeanmon commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/avm2_recursion_constraint.cpp:
```diff
@@ -52,6 +52,22 @@ void create_dummy_vkey_and_proof(Builder& builder,
     using Flavor = avm2::AvmFlavor;
 
+    // a lambda that sets dummy commitments
+    auto set_dummy_commitment = [&builder](const std::vector<stdlib::field_t<Builder>>& fields, size_t& offset) {
```
Very good idea to use this lambda!

jeanmon reviewed (commented):
LGTM.

fcarreiro commented on barretenberg/cpp/src/barretenberg/vm2/constraining/flavor.hpp:
```diff
@@ -266,27 +266,15 @@ class AvmFlavor {
         {
-            // TODO(https://github.com/AztecProtocol/barretenberg/issues/1466): Implement this function.
-            throw_or_abort("Not implemented yet!");
+            throw_or_abort("Not intended to be used because vk is hardcoded in circuit.");
```
Can we then remove this method completely? IIUC `Flavor` does not need to comply with any interface because everything is templated, and if someone were to call this it would just fail.

fcarreiro replied:
Or is the plan to add a constant here?

lucasxia01 replied:
No because it'll inherit the base class implementation. I just wanted to make sure that if we do ever accidentally call this, an error would be thrown

fcarreiro commented on barretenberg/cpp/src/barretenberg/vm2/constraining/recursion/recursive_flavor.hpp:
```diff
@@ -100,12 +98,25 @@ class AvmRecursiveFlavor {
                                   [[maybe_unused]] Transcript& transcript) const override
         {
-            throw_or_abort("Not implemented yet!");
+            throw_or_abort("Not intended to be used because vk is hardcoded in circuit.");
```
Same comment here.

LHerskind committed:
feat!: fiat-shamir AVM vk

Passes the vk to the prover and hardcodes the vk and vk hash in the recursive verifier. Adds logic to fiat-shamir VK hash in AVM, but stops short of doing it because padding makes things inconsistent in the recursive verifier. Also fixes NegativeBadPublicInputs test.
